test/Streamly/Test/Internal/Data/Parser.hs

```diff
@@ -0,0 +1,305 @@
+    forAll (listOf $ chooseInt (0, 10000)) $ \ls ->
+        case S.parse (P.all (> 5000)) (S.fromList ls) of
+            Right r -> r == (Prelude.all (> 5000) ls)
+            Left _ -> False
+
+yield :: Property
+yield = 
+    forAll (chooseInt (0, 10000)) $ \x ->
+        case S.parse (P.yield x) (S.fromList [1 :: Int]) of
+            Right r -> r == x
+            Left _ -> False
+
+yieldM :: Property
+yieldM =
+    forAll (chooseInt (0, 10000)) $ \x ->
+        case S.parse (P.yieldM $ return x) (S.fromList [1 :: Int]) of
+            Right r -> r == x
+            Left _ -> False
+
+die :: Property
+die =
+    property $
+    case S.parse (P.die "die test") (S.fromList [0 :: Int]) of
+        Right _ -> False
+        Left _ -> True
+
+dieM :: Property
+dieM =
+    property $
+    case S.parse (P.dieM (Right "die test")) (S.fromList [0 :: Int]) of
+        Right _ -> False
+        Left _ -> True
+
+-- Element Parser Tests
+
+peek :: Property
+peek = 
+    forAll (chooseInt (1, 100)) $ \list_length ->
+        forAll (vectorOf list_length (chooseInt (0, 10000))) $ \ls ->
+            case S.parse P.peek (S.fromList ls) of
+                Right head_value -> case ls of
+                    head_ls : _ -> head_value == head_ls
+                    _ -> False
+                Left _ -> False
+    .&&.
+    property (case S.parse P.peek (S.fromList []) of
+        Right _ -> False
+        Left _ -> True)
+
+eof :: Property
+eof = 
+    forAll (chooseInt (1, 100)) $ \list_length ->
+        forAll (vectorOf list_length (chooseInt (0, 10000))) $ \ls ->
+            case S.parse P.eof (S.fromList ls) of
+                Right _ -> False
+                Left _ -> True
+    .&&.
+    property (case S.parse P.eof (S.fromList []) of
+        Right _ -> True
+        Left _ -> False)
+
+satisfy :: Property
+satisfy = 
+    forAll (listOf (chooseInt (0, 10000))) $ \ls ->
+        case S.parse (P.satisfy predicate) (S.fromList ls) of
+            Right r -> case ls of
+                [] -> False
+                (x : _) -> predicate x && (r == x)
+            Left _ -> case ls of
+                [] -> True
+                (x : _) -> not $ predicate x
+        where
+            predicate = (>= 5000)
+
+-- Sequence Parsers Tests
+
+take :: Property
+take = 
+    forAll (chooseInt (0, 10000)) $ \n ->
+        forAll (listOf (chooseInt (0, 10000))) $ \ls ->
+            case S.parse (P.take n FL.toList) (S.fromList ls) of
+                Right parsed_list -> parsed_list == Prelude.take n ls
+                Left _ -> False
+
+takeEQ :: Property
+takeEQ =
+    forAll (chooseInt (0, 10000)) $ \n ->
+        forAll (listOf (chooseInt (0, 10000))) $ \ls ->
+            let 
+                list_length = Prelude.length ls
+            in
+                case S.parse (P.takeEQ n FL.toList) (S.fromList ls) of
+                    Right parsed_list -> (n <= list_length) && (parsed_list == Prelude.take n ls)
+                    Left _ -> n > list_length
```

Comment:
Separate a failure test to a separate property (n > list_length) and in one make property make sure at least `n` elements are always in the list and in another it need not be guaranteed.